### PR TITLE
Move creation of webpack classes to a cached factory

### PR DESCRIPTION
**Pull request by barak007**

### CODE REVIEW

**Comment by AviVahl** [comment]
make sure to increase the webpack `peerDependency` version to the minimum needed.

**Comment by barak007** [comment]
related to https://github.com/wix/stylable/issues/1710

File: packages/webpack-plugin/src/mini-css-support.ts [hunk 1]
```diff
@@ -1,23 +1,29 @@
...
     webpack: Compiler['webpack'],
     compilation: Compilation,
     staticPublicPath: string,
     stylableModules: Map<NormalModule, BuildData | null>,
     assetsModules: Map<string, NormalModule>
 ) {
-    const MiniCssExtractPlugin = getMiniCssExtractPlugin();
-    const CssModule = MiniCssExtractPlugin.getCssModule(webpack);
+    const MiniCssExtractPlugin = compilation.options.plugins.find(
+        (plugin) => plugin.constructor.name === 'MiniCssExtractPlugin'
```
Comment (AviVahl) [inline packages/webpack-plugin/src/mini-css-support.ts:15]: this assumes the plugins are all classes, but they might be objects with apply I think. not sure though. might wanna protect against constructor not being available.

File: packages/webpack-plugin/src/plugin-utils.ts [hunk 2]
```diff
@@ -151,8 +158,8 @@ export function extractDataUrlFromAssetModuleSource(source: string): string {
 type AssetNormalModule = NormalModule & { loaders: [{ loader: 'file-loader' | 'url-loader' }] };
 
 export function isLoadedWithKnownAssetLoader(module: Module): module is AssetNormalModule {
-    if (module instanceof NormalModule) {
-        return module.loaders.some(({ loader }) =>
+    if ('loaders' in module) {
+        return (module as any).loaders.some(({ loader }: { loader: string }) =>
```
Comment (AviVahl) [inline packages/webpack-plugin/src/plugin-utils.ts:162]: perhaps `as import('webpack').NormalModule`

File: packages/webpack-plugin/src/webpack-entities.ts [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+interface InjectDependencyTemplate {
+    new (
+        staticPublicPath: string,
+        stylableModules: Map<Module, BuildData | null>,
+        assetsModules: Map<string, NormalModule>,
+        runtimeStylesheetId: 'namespace' | 'module',
+        runtimeId: string,
+        cssInjection: 'js' | 'css' | 'mini-css' | 'none'
+    ): DependencyTemplate;
+    apply(
```
Comment (AviVahl) [inline packages/webpack-plugin/src/webpack-entities.ts:47]: remove. comes from dependencytemplate

File: packages/webpack-plugin/src/webpack-entities.ts [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+                source,
+                getReplacementToken('namespace'),
+                JSON.stringify(stylableBuildData.namespace)
+            );
+        }
+    }
+
+    class StylableRuntimeInject extends RuntimeModule {
+        constructor() {
+            super('stylable inject', 10);
```
Comment (AviVahl) [inline packages/webpack-plugin/src/webpack-entities.ts:253]: `RuntimeModule.STAGE_ATTACH` or whichever one is correct

File: packages/webpack-plugin/src/webpack-entities.ts [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+
+    class NoopTemplate {
+        apply() {
+            /** noop */
+        }
+    }
+
+    class StylableRuntimeStylesheet extends RuntimeModule {
+        constructor() {
+            super('stylable stylesheet', 10);
```
Comment (AviVahl) [inline packages/webpack-plugin/src/webpack-entities.ts:110]: RuntimeModule.STAGE_ATTACH

File: packages/webpack-plugin/src/mini-css-support.ts (hunk 1 above)
Comment (barak007) [inline packages/webpack-plugin/src/mini-css-support.ts:15]:
```suggestion
        (plugin) => plugin.constructor?.name === 'MiniCssExtractPlugin'
```

**Pull request merged**